rustls-libssl/src/callbacks.rs

```diff
@@ -0,0 +1,100 @@
+use core::cell::RefCell;
+use core::ffi::{c_uchar, c_void};
+use core::{ptr, slice};
+
+use openssl_sys::{SSL_TLSEXT_ERR_NOACK, SSL_TLSEXT_ERR_OK};
+
+use crate::entry::{SSL_CTX_alpn_select_cb_func, SSL};
+use crate::error::Error;
+
+/// Smuggling SSL* pointers from the outer entrypoint into the
+/// callback call site.
+pub struct SslCallbackContext;
+
+impl SslCallbackContext {
+    /// Register the original SSL* pointer for use in later callbacks.
+    ///
+    /// The returned object de-registers itself when dropped.
+    pub fn new(ssl: *mut SSL) -> Self {
+        SSL_CALLBACK_CONTEXT.set(Some(ssl));
+        Self
+    }
+
+    /// Get the original SSL* pointer, or else `NULL`
+    ///
+    /// This has thread-local semantics: it uses the most recent
+    /// object of this type created on this thread.
+    pub fn ssl_ptr() -> *mut SSL {
+        SSL_CALLBACK_CONTEXT.with_borrow(|holder| {
+            holder
+                .as_ref()
+                .map(|inner| *inner)
+                .unwrap_or_else(ptr::null_mut)
+        })
+    }
+}
+
+impl Drop for SslCallbackContext {
+    fn drop(&mut self) {
+        SSL_CALLBACK_CONTEXT.set(None);
+    }
+}
+
+thread_local! {
+    static SSL_CALLBACK_CONTEXT: RefCell<Option<*mut SSL>> = const { RefCell::new(None) };
+}
+
+/// Configuration needed to call [`invoke_alpn_callback`] later
+#[derive(Debug, Clone)]
+pub struct AlpnCallbackConfig {
+    pub cb: SSL_CTX_alpn_select_cb_func,
+    pub context: *mut c_void,
+}
+
+impl Default for AlpnCallbackConfig {
+    fn default() -> Self {
+        Self {
+            cb: None,
+            context: ptr::null_mut(),
+        }
+    }
+}
+
+/// Call a `SSL_CTX_alpn_select_cb_func` callback
+///
+/// Returns the selected ALPN, or None, or an error.
+pub fn invoke_alpn_callback(
+    config: &AlpnCallbackConfig,
+    offer: &[u8],
+) -> Result<Option<Vec<u8>>, Error> {
```

Comment:
Reading through these various callbacks I think the `invoke_xxxx_callback()` standalone fns might feel more natural expressed as fns on the callback types. Something like:

```rust
impl AlpnCallbackConfig {
  fn invoke(&self, offer: &[u8]) -> Result<Option<Vec<u8>>, Error> {
     ...
  }
}
```

Was there a reason to avoid that?

Reply:
Ah, that definitely looks nicer!